src/extensions/scratch3_boost/index.js

```diff
@@ -0,0 +1,1963 @@
+                            default: BoostColorLabel.BLACK,
+                            description: BoostColorLabel.BLACK
+                        }),
+                        value: BoostColorLabel.BLACK
+                    },
+                    {
+                        text: formatMessage({
+                            id: 'boost.color.any',
+                            default: BoostColorLabel.ANY,
+                            description: BoostColorLabel.ANY
+                        }),
+                        value: BoostColorLabel.ANY
+                    }
+                ],
+                OP: ['<', '>']
+            }
+        };
+    }
+
+    /**
+     * Turn specified motor(s) on for a specified duration.
+     * @param {object} args - the block's arguments.
+     * @property {MotorID} MOTOR_ID - the motor(s) to activate.
+     * @property {int} DURATION - the amount of time to run the motors.
+     * @return {Promise} - a promise which will resolve at the end of the duration.
+     */
+    motorOnFor (args) {
+        // TODO: cast args.MOTOR_ID?
+        let durationMS = Cast.toNumber(args.DURATION) * 1000;
+        durationMS = MathUtil.clamp(durationMS, 0, 15000);
+        return new Promise(resolve => {
+            this._forEachMotor(args.MOTOR_ID, motorIndex => {
+                const motor = this._peripheral.motor(motorIndex);
+                if (motor) {
+                    motor.turnOnFor(durationMS);
+                }
+            });
+
+            // Run for some time even when no motor is connected
+            setTimeout(resolve, durationMS);
+        });
+    }
+
+    /**
+     * Turn specified motor(s) on for a specified rotation in full rotations.
+     * @param {object} args - the block's arguments.
+     * @property {MotorID} MOTOR_ID - the motor(s) to activate.
+     * @property {int} ROTATION - the amount of full rotations to turn the motors.
+     * @return {Promise} - a promise which will resolve at the end of the duration.
+     */
+    motorOnForRotation (args) {
+        // TODO: cast args.MOTOR_ID?
+        let degrees = Cast.toNumber(args.ROTATION) * 360;
+        // TODO: Clamps to 100 rotations. Consider changing.
+        const sign = Math.sign(degrees);
+        degrees = Math.abs(MathUtil.clamp(degrees, -360000, 360000));
+
+        const motors = [];
+        this._forEachMotor(args.MOTOR_ID, motorIndex => {
+            motors.push(motorIndex);
+        });
+
+        const promises = motors.map(portID => {
+            const motor = this._peripheral.motor(portID);
+            if (motor) {
+                return new Promise(resolve => {
+                    motor.turnOnForDegrees(degrees, sign);
+                    motor.pendingPromiseFunction = resolve;
+                });
+            }
+            return null;
+        });
+        // To prevent the block from returning a value, an empty function is added to the .then
+        return Promise.all(promises).then(() => {});
```

Comment:
It was added because without it, when running motor-commands on all ports without motors plugged into those ports, the block commands for the non-motor ports would execute immediately. (hope that made sense, it's hard to describe 😆
By mapping only the ports that have motors on them, and checking the promises against `Promise.all`, the blocks yield correctly. 

I do agree that it doesn't read very well. so I'll try to add some comments to the code!